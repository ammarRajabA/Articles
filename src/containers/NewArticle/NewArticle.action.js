import fakeArticles from '../../fakeAPI/fake_articles'
import history from '../../routes/history'
import articles from '../../fakeAPI/articles'

export const saveArticle=(article)=>dispatch=>{
	dispatch({type:"START_POST_ARTICLE"})
	// Use Axios or Fetch to POST article to backend
	fakeArticles.push(article)
	dispatch({type:"FINISH_POST_ARTICLE"});
	history.push('/articles');
}

export const updateArticle=(id,article)=>dispatch=>{
	dispatch({type:"START_PATCH_ARTICLE"})
	// Use Axios or Fetch to POST article to backend
	articles.patch(id,article)
	dispatch({type:"FINISH_PATCH_ARTICLE"});
	history.push('/articles');
}