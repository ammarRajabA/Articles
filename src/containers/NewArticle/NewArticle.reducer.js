const INIT={
	loading:false
}

export default (state=INIT,action)=>{
	switch(action.type){
		case 'START_POST_ARTICLES':
			return {...state,loading:true}
		case 'FINISH_POST_ARTICLES':
			return {...state,loading:false}
		case 'START_PATCH_ARTICLES':
			return {...state,loading:true}
		case 'FINISH_PATCH_ARTICLES':
			return {...state,loading:false}
		default:
			return state;
	}
}