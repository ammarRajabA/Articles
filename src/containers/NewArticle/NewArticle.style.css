.new-article-container{
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: center;
	margin-right: auto;
	margin-left: auto;
	width: 80%;
	margin-top: 20px;
}
.new-article-container .cover-image-container{
	position: relative;
	width: 400px;
	height: 300px;
	transition: 0.3s;
	margin-bottom: 20px;
}
.new-article-container .cover-image-container:hover{
	opacity: 0.7;
}
.new-article-container .cover-image-container input{
	display: none;
}
.new-article-container .cover-image-container .error{
	position: absolute;
	top: 0px;
	right: 0px;
	width: 100%;
	background-color: #DD5555;
	padding-bottom: 2px;
	padding-top: 2px;
	padding-right: 10px;
	padding-left: 10px;
}
.new-article-container .cover-image-container .label{
	position: absolute;
	bottom: 10px;
	right: 10px;
	color: white;
	font-size: 12px;
	background-color: #555555;
	padding-bottom: 2px;
	padding-top: 2px;
	padding-right: 10px;
	padding-left: 10px;
	border-radius: 5px;
	cursor: pointer;
}
.new-article-container .cover-image-container .cover-image{
	width: 100%;
	height: 100%;
	margin-bottom: 10px;
}
.new-article-container .editor-container{
	margin-top: 20px;
	margin-bottom: 20px;
}
.editor-container .editor-wrapper{
	border: 1px solid #E3E3E3;
}