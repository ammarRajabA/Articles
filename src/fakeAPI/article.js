import articles from './fake_articles'
export default {
	get:(id)=>{
		return articles.find((article)=>article.id===id)
	},
	delete:(id)=>{
		var target=articles.findIndex((article)=>article.id===id);
		articles.splice(target,1)
	}
}