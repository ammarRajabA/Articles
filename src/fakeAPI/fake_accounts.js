export default[
	{
		firstName:'Ammar',
		lastName:'Rajab',
		email:'ammar.rajab1@gmail.com',
		password:'12345',
		id:'12345'
	}
]