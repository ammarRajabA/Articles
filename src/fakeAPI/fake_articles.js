export default[
	{
		title:"first",
		content:JSON.stringify({"blocks":[{"key":"5viej","text":"This is my first article, it so cool I can color words ","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":49,"length":5,"style":"color-rgb(243,121,52)"}],"entityRanges":[],"data":{}},{"key":"dh0hv","text":"I can also make bold italic or any other style.","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":16,"length":5,"style":"BOLD"},{"offset":21,"length":7,"style":"ITALIC"}],"entityRanges":[],"data":{}},{"key":"blt1q","text":"I also can write math 25 =32","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":23,"length":2,"style":"SUPERSCRIPT"}],"entityRanges":[],"data":{}},{"key":"ddclt","text":"ðŸ“£ Attention here : My Github account Ammar Rajab ","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":2,"length":48,"style":"color-rgb(209,72,65)"}],"entityRanges":[{"offset":38,"length":11,"key":0}],"data":{}}],"entityMap":{"0":{"type":"LINK","mutability":"MUTABLE","data":{"url":"https://github.com/ammarrajaba","targetOption":"_blank"}}}}),
		image:"https://picsum.photos/id/258/200/300",
		author:"Ammar Rajab",
		id:"13245"
	},
	{
		title:"Second",
		content:JSON.stringify({"blocks":[{"key":"35tp4","text":"Another article by me,","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{}},{"key":"a1v72","text":"It shows how we align text","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{}},{"key":"cifh6","text":"Left","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{}},{"key":"b5gsb","text":"Middle","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{"text-align":"center"}},{"key":"eopjr","text":"and right","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{"text-align":"right"}},{"key":"1h6bs","text":"","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{}},{"key":"d3mrc","text":"and change font","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":0,"length":15,"style":"fontfamily-Impact"}],"entityRanges":[],"data":{}}],"entityMap":{}}),
		image:"https://picsum.photos/id/232/200/300",
		author:"Ammar Rajab",
		id:"15555"
	},
	{
		title:"Third",
		content:JSON.stringify({"blocks":[{"key":"5viej","text":"This is my first article, it so cool I can color words ","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":49,"length":5,"style":"color-rgb(243,121,52)"}],"entityRanges":[],"data":{}},{"key":"dh0hv","text":"I can also make bold italic or any other style.","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":16,"length":5,"style":"BOLD"},{"offset":21,"length":7,"style":"ITALIC"}],"entityRanges":[],"data":{}},{"key":"blt1q","text":"I also can write math 25 =32","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":23,"length":2,"style":"SUPERSCRIPT"}],"entityRanges":[],"data":{}},{"key":"ddclt","text":"ðŸ“£ Attention here : My Github account Ammar Rajab ","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":2,"length":48,"style":"color-rgb(209,72,65)"}],"entityRanges":[{"offset":38,"length":11,"key":0}],"data":{}}],"entityMap":{"0":{"type":"LINK","mutability":"MUTABLE","data":{"url":"https://github.com/ammarrajaba","targetOption":"_blank"}}}}),
		image:"https://picsum.photos/id/258/200/300",
		author:"Ammar Rajab",
		id:"13445"
	},
	{
		title:"Forth",
		content:JSON.stringify({"blocks":[{"key":"35tp4","text":"Another article by me,","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{}},{"key":"a1v72","text":"It shows how we align text","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{}},{"key":"cifh6","text":"Left","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{}},{"key":"b5gsb","text":"Middle","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{"text-align":"center"}},{"key":"eopjr","text":"and right","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{"text-align":"right"}},{"key":"1h6bs","text":"","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{}},{"key":"d3mrc","text":"and change font","type":"unstyled","depth":0,"inlineStyleRanges":[{"offset":0,"length":15,"style":"fontfamily-Impact"}],"entityRanges":[],"data":{}}],"entityMap":{}}),
		image:"https://picsum.photos/id/232/200/300",
		author:"Ammar Rajab",
		id:"15005"
	}
]
