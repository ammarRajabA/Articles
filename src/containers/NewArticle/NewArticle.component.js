import React, {Component} from 'react';
import { connect } from 'react-redux';
import { Button, Form, FormGroup, Label, Input, FormFeedback } from 'reactstrap';
import {Redirect} from 'react-router-dom'

import { Editor } from 'react-draft-wysiwyg';
import { EditorState, convertToRaw, convertFromRaw} from 'draft-js';

import {saveArticle,updateArticle} from './NewArticle.action';

import {genID} from '../../helpers/random';
import history from '../../routes/history'

import 'react-draft-wysiwyg/dist/react-draft-wysiwyg.css';
import './NewArticle.style.css';

export class NewArticle extends Component{
	constructor(props){
		super(props);
		var newEditorState= EditorState.createEmpty();
		this.user=JSON.parse(localStorage.getItem('user'));
		if (this.user)
			this.state={
				showMessage:false,
				message:"",
				title:"",
				editorState: newEditorState,
				author:this.user.firstName+' '+this.user.lastName,
				image:'',
				id:genID(),
				newArticle:true
			}
	}
	componentDidMount(){
		if (this.props.match.params.id){
			if (this.user){
				var newEditorState=EditorState.createWithContent(convertFromRaw(JSON.parse(this.props.article.content)))
				this.setState({
					title:this.props.article.title,
					editorState:newEditorState,
					author:this.user.firstName+' '+this.user.lastName,
					image:this.props.article.image,
					id:this.props.article.id,
					newArticle:false
				})
			}
		}
	}
	componentDidUpdate(prevProps){
		if (this.props.match.params.id !== prevProps.match.params.id){
			if (this.user){
				var newEditorState=EditorState.createWithContent(convertFromRaw(JSON.parse(this.props.article.content)))
				this.setState({
					title:this.props.article.title,
					editorState:newEditorState,
					author:this.user.firstName+' '+this.user.lastName,
					image:this.props.article.image,
					id:this.props.article.id,
					newArticle:false
				})
			}
		}
	}
	handleUpload=(e)=>{
		var file=e.target.files[0];
		if (file.size>(4*1024*1024)){
			this.setState({showMessage:true,message:"Maximum file size is 4 MB, this file is "+((e.target.files[0].size/1024)/1024).toFixed(2)+" MB"},()=>{
				setTimeout(()=>this.setState({showMessage:false,message:""}),5000)
			})
		}else{
			var reader  = new FileReader();
			reader.addEventListener("load", ()=> {
				this.setState({image:reader.result})
			}, false);
			reader.readAsDataURL(file);
		}
	}
	renderCoverImage=()=>{
		return (
				<span className="cover-image-container">
					<img src={this.state.image} className="cover-image"/>
					<label htmlFor={"image"} className="label">Click here to upload image</label>
					<Input type="file" accept="image/png, image/jpeg" id={"image"} onChange={this.handleUpload}/>
					{this.state.showMessage&&<label className="error">{this.state.message}</label>}
				</span>
			)
	}
	renderForm=()=>{
		return(
				<div className="form-container">
					<Form>
						<FormGroup>
							<Input placeholder="Title" value={this.state.title} onChange={(e)=>this.setState({title:e.target.value})}/>
						</FormGroup>
					</Form>
					<Button size="lg" onClick={this.handleSubmit}>Submit Article</Button>
				</div>
			)
	}
	renderEditor=()=>{
		// debug for Testing since testing Editor causes test to freeze
		if (this.props.debug) return <div></div>
		return(
				<div  className="editor-container">
					<Editor
						wrapperClassName="editor-wrapper"
						editorState={this.state.editorState}
						onEditorStateChange={(editorState)=>this.setState({editorState})}
					/>
				</div>
			)
	}
	handleSubmit=()=>{
		var editorState=JSON.stringify(convertToRaw(this.state.editorState.getCurrentContent()))
		if (this.state.newArticle)
			this.props.saveArticle({
				title:this.state.title,
				content:editorState,
				author:this.state.author,
				image:this.state.image,
				id:this.state.id
			})
		else
			this.props.updateArticle(this.state.id,{
				title:this.state.title,
				content:editorState,
				author:this.state.author,
				image:this.state.image,
				id:this.state.id
			})
	}
	render(){
		if (!this.user) return <Redirect to="/"/>
		return (
			<div className="new-article-container">
				{this.renderCoverImage()}
				{this.renderForm()}
				{this.renderEditor()}
			</div>
		)
	}
}

const mapStateToProps=(state)=>{
	return {article:state.article}
}

export default connect(mapStateToProps,{saveArticle,updateArticle})(NewArticle);