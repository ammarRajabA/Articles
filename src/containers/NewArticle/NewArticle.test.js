import React from 'react'
import { shallow ,mount } from 'enzyme';

import {NewArticle} from './NewArticle.component';
import fake_articles from '../../fakeAPI/fake_articles';
import fake_accounts from '../../fakeAPI/fake_accounts';

import article from '../../fakeAPI/article'

class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
};

global.localStorage = new LocalStorageMock;

const test_props={
	match:{
		params:{
			id:"13245"
		}
	},
	article:{
		...article.get("13245"),
		loading:false
	}
}

describe('Article component',function () {

	it('match snapshot',function () {
		localStorage.setItem("user",JSON.stringify(fake_accounts[0]))
		const component = mount(
			<NewArticle 
				match={test_props.match} 
				article={test_props.article}
				debug
			/>);
		
		expect(component).toMatchSnapshot();
		localStorage.clear()
	})
})