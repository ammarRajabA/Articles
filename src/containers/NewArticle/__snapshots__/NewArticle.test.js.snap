// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article component match snapshot 1`] = `
<NewArticle
  article={
    Object {
      "author": "Ammar Rajab",
      "content": "{\\"blocks\\":[{\\"key\\":\\"5viej\\",\\"text\\":\\"This is my first article, it so cool I can color words \\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[{\\"offset\\":49,\\"length\\":5,\\"style\\":\\"color-rgb(243,121,52)\\"}],\\"entityRanges\\":[],\\"data\\":{}},{\\"key\\":\\"dh0hv\\",\\"text\\":\\"I can also make bold italic or any other style.\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[{\\"offset\\":16,\\"length\\":5,\\"style\\":\\"BOLD\\"},{\\"offset\\":21,\\"length\\":7,\\"style\\":\\"ITALIC\\"}],\\"entityRanges\\":[],\\"data\\":{}},{\\"key\\":\\"blt1q\\",\\"text\\":\\"I also can write math 25 =32\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[{\\"offset\\":23,\\"length\\":2,\\"style\\":\\"SUPERSCRIPT\\"}],\\"entityRanges\\":[],\\"data\\":{}},{\\"key\\":\\"ddclt\\",\\"text\\":\\"ðŸ“£ Attention here : My Github account Ammar Rajab \\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[{\\"offset\\":2,\\"length\\":48,\\"style\\":\\"color-rgb(209,72,65)\\"}],\\"entityRanges\\":[{\\"offset\\":38,\\"length\\":11,\\"key\\":0}],\\"data\\":{}}],\\"entityMap\\":{\\"0\\":{\\"type\\":\\"LINK\\",\\"mutability\\":\\"MUTABLE\\",\\"data\\":{\\"url\\":\\"https://github.com/ammarrajaba\\",\\"targetOption\\":\\"_blank\\"}}}}",
      "id": "13245",
      "image": "https://picsum.photos/id/258/200/300",
      "loading": false,
      "title": "first",
    }
  }
  debug={true}
  match={
    Object {
      "params": Object {
        "id": "13245",
      },
    }
  }
>
  <div
    className="new-article-container"
  >
    <span
      className="cover-image-container"
    >
      <img
        className="cover-image"
        src="https://picsum.photos/id/258/200/300"
      />
      <label
        className="label"
        htmlFor="image"
      >
        Click here to upload image
      </label>
      <Input
        accept="image/png, image/jpeg"
        id="image"
        onChange={[Function]}
        type="file"
      >
        <input
          accept="image/png, image/jpeg"
          className="form-control-file"
          id="image"
          onChange={[Function]}
          type="file"
        />
      </Input>
    </span>
    <div
      className="form-container"
    >
      <Form
        tag="form"
      >
        <form
          className=""
        >
          <FormGroup
            tag="div"
          >
            <div
              className="form-group"
            >
              <Input
                onChange={[Function]}
                placeholder="Title"
                type="text"
                value="first"
              >
                <input
                  className="form-control"
                  onChange={[Function]}
                  placeholder="Title"
                  type="text"
                  value="first"
                />
              </Input>
            </div>
          </FormGroup>
        </form>
      </Form>
      <Button
        color="secondary"
        onClick={[Function]}
        size="lg"
        tag="button"
      >
        <button
          aria-label={null}
          className="btn btn-secondary btn-lg"
          onClick={[Function]}
          type="button"
        >
          Submit Article
        </button>
      </Button>
    </div>
    <div />
  </div>
</NewArticle>
`;
